### PR TITLE
feat(operator): add switchFirst and switchMapFirst

### PR DESCRIPTION
**Pull request by jinroh**
(Relates to #645)

Adds `switchFirst` and `switchMapFirst` operators (names still in discussion in #645) with tests and micro perfs.

Tell me if anything is missing from your guidelines !

### CODE REVIEW

**Comment by jinroh** [comment]
Oups.. yes that was my intention ! I didn't find the interface because it is inline is the `Rx.KitchenSink.ts` file.

PR updated and rebased.

File: src/Rx.KitchenSink.ts [hunk 1]
```diff
@@ -13,6 +13,8 @@ interface KitchenSinkOperators<T> extends CoreOperators<T> {
   min?: <T, R>(comparer?: (x: R, y: T) => R) => Observable<R>;
   timeInterval?: <T>(scheduler?: IScheduler) => Observable<T>;
   mergeScan?: <T, R>(project: (acc: R, x: T) => Observable<R>, seed: R) => Observable<R>;
+  switchFirst?: <R>() => Observable<R>
```
Comment (kwonoj) [inline src/Rx.KitchenSink.ts]: Is type signature for `switchFirst` intended to be as below?
```
Observable<T> {
 switchFirst<R>:Observable<R> {
 }
}
```

In `switchFirst.ts`, it seems as
```
Observable<T> {
 switchFirst(): Observable<T> {
 }
}
```

File: spec/operators/switchFirst-spec.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+var Observable = Rx.Observable;
+var immediateScheduler = Rx.Scheduler.immediate;
+
+describe('Observable.prototype.switchFirst()', function () {
+  it('should switch to first immediately-scheduled inner Observable', function () {
+    var e1 = cold( '(ab|)');
+    var e2 = cold( '(cd|)');
+    var expected = '(ab|)';
+    expectObservable(Observable.of(e1, e2).switchFirst()).toBe(expected);
+  });
```
Comment (staltz) [inline spec/operators/switchFirst-spec.js]: Can you add subscription tests also for these basic cases? Sometimes we discover bugs related to synchronous subscribe and unsubscribe `expected: (^!)` and sometimes unsubscription never happens (and should).

File: spec/operators/switchFirst-spec.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    var xsubs =    '      ^         !           ';
+    var y = cold(                '---d--e---f---|');
+    var ysubs = [];
+    var e1 = hot(  '------x-------y------|       ', { x: x, y: y });
+    var unsub =    '                !            ';
+    var expected = '--------a---b---             ';
+    expectObservable(e1.switchFirst(), unsub).toBe(expected);
+    expectSubscriptions(x.subscriptions).toBe(xsubs);
+    expectSubscriptions(y.subscriptions).toBe(ysubs);
+  });
```
Comment (staltz) [inline spec/operators/switchFirst-spec.js]: Perfect. :+1:

File: spec/operators/switchMapFirst-spec.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+    var e1 = Observable.throw('damn');
+    var expected = '#';
+    expectObservable(e1.switchMapFirst(function (a) { return Observable.of(a); })).toBe(expected, null, 'damn');
+  });
+
+  it('should handle outer empty', function () {
+    var e1 = Observable.empty();
+    var expected = '|';
+    expectObservable(e1.switchMapFirst(function (a) { return Observable.of(a); })).toBe(expected);
+  });
```
Comment (staltz) [inline spec/operators/switchMapFirst-spec.js]: Also these cases could get subscription tests.

File: spec/operators/switchFirst-spec.js (hunk 2 above)
Comment (jinroh) [inline spec/operators/switchFirst-spec.js]: Of course ! I'll do that over the weekend or next week.

File: src/Rx.KitchenSink.ts (hunk 1 above)
Comment (jinroh) [inline src/Rx.KitchenSink.ts]: Indeed. I fixed it.

The `switch` operator is also wrong, I'll do a small PR to fix that. Thanks !

**Comment by jinroh** [comment]
Added some more subscription tests as @staltz suggested and rebased.

**Comment by jinroh** [comment]
Micro perf results:

```
                                         |                     RxJS 4.0.6 |             RxJS 5.0.0-alpha.8 |          factor |      % improved
---------------------------------------------------------------------------------------------------------------------------------------------------
              switchMapFirst - immediate |                 3,094 (±0.73%) |                14,552 (±1.09%) |           4.70x |          370.4%
                          switchMapFirst |                13,568 (±0.51%) |               145,489 (±0.47%) |          10.72x |          972.3%
                 switchFirst - immediate |                27,101 (±0.58%) |                12,298 (±0.59%) |           0.45x |          -54.6%
                             switchFirst |                17,275 (±0.52%) |               120,344 (±0.38%) |           6.97x |          596.6%
```

Can't really explain the `switchFirst - immediate` results... Are you measuring the same things ?

File: src/Rx.KitchenSink.ts (hunk 1 above)
Comment (jinroh) [inline src/Rx.KitchenSink.ts]: Just posted #669 to fix that.

**Comment by jinroh** [comment]
Forget it, I copy/pasted my perf benchs from a broken one :innocent:.
Fixed the original benchmarks in #670.

File: src/Rx.KitchenSink.ts [hunk 5]
```diff
@@ -13,6 +13,8 @@ interface KitchenSinkOperators<T> extends CoreOperators<T> {
   min?: <T, R>(comparer?: (x: R, y: T) => R) => Observable<R>;
   timeInterval?: <T>(scheduler?: IScheduler) => Observable<T>;
   mergeScan?: <T, R>(project: (acc: R, x: T) => Observable<R>, seed: R) => Observable<R>;
+  switchFirst?: () => Observable<T>
```
Comment (kwonoj) [inline src/Rx.KitchenSink.ts]: missing `;`? ;)

File: src/operators/switchMapFirst.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import OuterSubscriber from '../OuterSubscriber';
+import subscribeToResult from '../util/subscribeToResult';
+
+export default function switchMapFirst<T, R, R2>(project: (value: T, index: number) => Observable<R>,
+                                                 resultSelector?: (outerValue: T,
+                                                                   innerValue: R,
+                                                                   outerIndex: number,
+                                                                   innerIndex: number) => R2): Observable<R> {
+  return this.lift(new SwitchMapFirstOperator(project, resultSelector));
+}
```
Comment (kwonoj) [inline src/operators/switchMapFirst.ts]: Seems this declaration can be updated to remove `R2` comparing to operator declaration : 
`switchMapFirst?: <R>(project: ((x: T, ix: number) => Observable<any>), projectResult?: (x: T, y: any, ix: number, iy: number) => R) => Observable<R>;`

**Comment by blesh** [comment]
@jinroh this looks good, but there are conflicts to be resolved, can you rebase?

I'm not sure about the names, yet, but we can bike-shed those at another time. The work looks great.

**Comment by jinroh** [comment]
@blesh done !

**Comment by blesh** [comment]
Thanks @jinroh! Also you might be proud to know you've pushed us up over 1,500 unit tests!

**Pull request merged**